### DISCUSSION TITLE
Customized colors.

### DISCUSSION THREAD

**Discussion by pml-lang** [tajmone/Sublime-PML#7] [Ideas]
According to questions [sublime text 3 - change syntax colours](https://stackoverflow.com/questions/17824568/sublime-text-3-change-syntax-colours) and [Sublime Text 3: Customize certain syntax highlighting](https://superuser.com/questions/1336368/sublime-text-3-customize-certain-syntax-highlighting), changing a specific color in Sublime Text is cumbersome, especially for non-technical people.

I wonder if there is an easier way to do it, or maybe a plugin.

Ideas (for later):

- Ideally, the user would be able to do that in a GUI.
- Alternatively, there could be a CSS file (don't know if ST supports that).
- Or there could be a config file with a simple format like this:
```
tag_color = green
attribute_name_color = rgb(0,0,255)
attribute_value_color = #FAFAFA
error_color = red, yellow  // foreground, background
```

If this is not possible (or to complicated) it would be good to have a small description for customizing colors in the README file, or maybe just a link to online instructions.

**Comment by tajmone** [comment]
> Ideas (for later):
> 
> - Ideally, the user would be able to do that in a GUI.
> - Alternatively, there could be a CSS file (don't know if ST supports that).
> - Or there could be a config file with a simple format like this:

All of the above might be quite hard to implement. There are some packages which deal specifically with exposing color schemes to end users in a simplified manner, so end users might be better off looking for similar packages instead.

With the newer color scheme format for ST3 it's not all that complicate, it's just a JSON file which allows to customize elements' color, text styles, and other aspects. The older scheme format was in XML format, and a bit more complex to handle than the newer one.

While working on the PML syntax I'm testing how it looks like with various schemes (either native or popular ones), and of course each scheme is different, ignoring some elements and/or focusing more on others. As a general rule, if we stick to the scoping guidelines, markup elements should look OK on all schemes (e.g. italics in italic, bold in bold, etc.).

Good schemes are designed to cover a broad range of programming languages and syntaxes, including markup syntaxes; this type of coverage is the hardest part for it requires experience with the various languages.

In some ST syntaxes that I created in the past, I also included in the package one or two default schemes targeting the language of the package, so we could also do that in this package. Many markup syntax packages do this, e.g. the Markdown ST package ships with five schemes designed especially for editing markdown, with either a light or dark theme.

### Hashed Syntax Highlighting

- [Sublime Text 3 Documentation » Color Schemes]:
    + [Hashed Syntax Highlighting]

I wouldn't say that creating a ST scheme is hard, it's just a bit tedious and time consuming.

I was thinking that for a custom PML color scheme it would be nice to use the [Hashed Syntax Highlighting] feature, where:

> each token matching the scope specified will receive a unique color from one, or more, gradients. Some editors refer to this style of highlighting as "Semantic Highlighting".

I did experiment with it, and I think it's ideal for a syntax like PML, where you have a lot of nested nodes, for it allows to clearly distinguish which closing bracket is associated with which opening tag, since each node would then get a slightly different color from the other nodes. If one doesn't overdo this, i.e. avoiding the scheme to look like a Christmas tree, it can be pleasant to work with — usually picking a gradient of some temperature colors (e.g. from yellow to red) works best, for it prevents having too many colors.





[Sublime Text 3 Documentation » Color Schemes]: https://www.sublimetext.com/docs/3/color_schemes.html

[Hashed Syntax Highlighting]: https://www.sublimetext.com/docs/3/color_schemes.html#hashed_syntax_highlighting

**Reply by pml-lang** [reply]
> included in the package one or two default schemes targeting the language of the package

That would surely be an excellent add-on in the final package (the cherry on the cake). I imagine it's probably impossible to provide standard-out-of-the-box colors that work well for all people in all schemes.

> Semantic Highlighting

Very interesting. Thanks for the link. I'm sure you'll find the right balance to make it useful without overdoing.
